README.md
# SeoCache

[![Gem Version](https://badge.fury.io/rb/seo_cache.svg)](https://badge.fury.io/rb/seo_cache)
[![Ruby Style Guide](https://img.shields.io/badge/code_style-rubocop-brightgreen.svg)](https://github.com/rubocop-hq/rubocop)
[![Build Status](https://travis-ci.org/floXcoder/seo_cache.svg)](https://travis-ci.org/floXcoder/seo_cache)

**SeoCache is a dedicated cache for SEO with JavaScript rendering** :fire:

You can find in this article, a more detail explanation of the purpose and usage of this gem:

- In english: https://www.ginkonote.com/users/flo/articles/seo-generate-and-cache-your-pages@seo
- In french: https://www.ginkonote.com/fr/utilisateurs/flo/articles/seo-generation-et-mise-en-cache-des-pages@seo

## Table of contents
- [Purpose](#purpose)
- [Installation](#installation)
- [Configuration](#configuration)
- [Helpers for controllers](#helpers-for-controllers)
- [Check cached pages](#check-cached-pages)
- [Automatic caching](#automatic-caching)
- [Server configuration](#server-configuration)
- [Inspiration](#inspiration)
- [Contributing](#contributing)

## Purpose

Google credo is: Don't waste my bot time!

In order to reduce Googlebot crawling time, let's provide the complete HTML files in a specific cache.

This cache is suitable for static (generated or not) pages but not for user private pages.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'seo_cache'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install seo_cache

Install chromium or chrome driver on your device (the chromedriver will be automatically downloaded).

Declare the middleware. For instance in `config/initializers/seo_cache.rb`:

```ruby
require 'seo_cache'

# See options below

Rails.application.config.middleware.use SeoCache::Middleware
```

Required at least Ruby 2.6.

## Configuration

Chrome path (**required**) (`disk` or `memory`):

```ruby
SeoCache.chrome_path = Rails.env.development? ? '/usr/bin/chromium-browser' : '/usr/bin/chromium'
```

Choose a cache mode (`memory` (default) or `disk`):

```ruby
SeoCache.cache_mode = 'memory'
```

Cache path (**required**):
    
```ruby
SeoCache.cache_path = Rails.root.join('public', 'seo_cache')
```

Redis URL (**required** if memory cache):

```ruby
SeoCache.redis_url = "redis://localhost:6379/"
```

Redis prefix (if needed):

```ruby
SeoCache.redis_namespace = '_my_project:seo_cache'
```

Redis database index (if needed):

```ruby
SeoCache.redis_db_index = 1
```

Waiting time before the page is fully loaded (optional in sec, useful for async elements):
    
```ruby
SeoCache.wait_time_for_page_loading = 3
```

Specific log file (if you want to log missed cache urls):
    
```ruby
SeoCache.logger_path = Rails.root.join('log', 'seo_cache.log')
```

Activate missed cache urls:
    
```ruby
SeoCache.log_missed_cache = true
```
 
Domains to whitelist (authorize only domains which contains these hosts):

```ruby
SeoCache.whitelist_hosts = []
```

URLs to blacklist:

```ruby
SeoCache.blacklist_urls = %w[^/assets/.* ^/admin.*]
```

Params to blacklist:

```ruby
SeoCache.blacklist_params = %w[page]
```

URLs to whitelist:

```ruby
SeoCache.whitelist_urls = []
```

Cache page even if user is connected (be careful when caching sensitive data):

```ruby
SeoCache.cache_with_user_connected = true
```

Parameter to add manually to the URl to force page caching, if you want to cache a specific URL (e.g. `https://<my_website>/?_seo_cache_=true`):

```ruby
SeoCache.force_cache_url_param = '_seo_cache_'
```

Cache only the pages with these HTTP status code (don't cache by default not found or error pages):

```ruby
SeoCache.cache_only_status = [<your_list>]
```

URL extension to ignore when caching (already defined):

```ruby
SeoCache.extensions_to_ignore = [<your_list>]
```

List of bot agents (already defined):

```ruby
SeoCache.crawler_user_agents = [<your_list>]
```
    
Parameter added to URL when generating the page, avoid infinite rendering (override only if already used):

```ruby
SeoCache.prerender_url_param = '_prerender_'
```

If you encounter the following error `DevToolsActivePort file doesn't exist`, you can add the following option:

```ruby
SeoCache.chrome_debugging_port = '9222'
```

Be aware, JS will be rendered twice: once by this gem and once by client. For React, this not a problem but with jQuery plugins, it can duplicate elements in the page (you have to check the redundancy).

Disk cache is recommended by default. Nginx will directly fetch HTML files on disk. The TTFB (time to first byte) will be under 200ms :). You can use memory cache if you have lot of RAM, but if you shut down your server, you will lost all the generated pages! So prefer cache disk storage.

## Helpers for controllers

You can check if seo mode is active in your controllers, with the following variable:

```ruby
def check_seo_mode
  @seo_mode = (params.key?(SeoCache.prerender_url_param) || params.key?(SeoCache.force_cache_url_param))
end
```

And if you want to access to this variable in JS files:

```javascript
window.seoMode = "{@seo_mode}"
```


## Check cached pages

Too see in browser the cached page, open a browser and set the user agent to:

`Googlebot (Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html))`

For instance, with Chrome or Chromium, you can change the user agent in "Network" / "Network conditions" panel.

To (re)cache a page, add this parameter to the url (and the browser must use the Googlebot user agent):

`/?_seo_cache_=true`

## Automatic caching

To automate caching, create a cron rake task (e.g. in `lib/tasks/populate_seo_cache.rake`):

```ruby
namespace :my_project do

  desc 'Populate cache for SEO'
  task populate_seo_cache: :environment do |_task, _args|
    require 'seo_cache/populate_cache'
    
    paths_to_cache = public_paths_like_sitemap
    
    SeoCache::PopulateCache.new('https://<your-domain-name>', paths_to_cache).new.perform
  end
end
```

You can add the `force_cache: true` option to `SeoCache::PopulateCache` for overwrite cached data.

If your website has several domains (.com, .fr, ...), you need to generate your urls for each locale (each key is the locale). And use the `with_locale_keys: true` option.

If you want to execute only through a rake task, you can comment the line which include the middleware. keep all options configured and remove only the middleware. Thus all pages will be cached and SeoCache isn't called for pages not in cache.
It's useful if you have a script which generates all website pages (based on sitemap for instance) and you run script every day.

## Server configuration

If you use disk caching, add this to your Nginx configuration:

```nginx
# Before any block:
map $http_user_agent $limit_bots {
    default 0;
    ~*(google|bing|yandex|msnbot) 1;
}

map $request_method $is_get {
    default 0;
    GET 1;
}

# Before location block:
set $bot_request '';
if ($is_get) {
      set $bot_request 'GET_';
}

if ($limit_bots) { 
  set $bot_request "${bot_request}BOT"; 
}

set $cache_extension '';
if ($bot_request = GET_BOT) {
  set $cache_extension '.html';
}

# Inside location block:
location / {
    # Ignore url with blacklisted params (e.g. page)
    if ($arg_page) {
      break;
    }
    if ($arg__seo_cache_) {
      break;
    }

    # Local rewrite to remove trailing slash
    rewrite ^/(.*)/$ /$1 last;
    
    try_files /seo_cache/$uri/index$cache_extension /seo_cache/$uri$cache_extension /seo_cache/$uri $uri @rubyproxy;
    # Or for multi-domains:
    # try_files /seo_cache/fr/$uri/index$cache_extension /seo_cache/fr/$uri$cache_extension /seo_cache/fr/$uri $uri @rubyproxy;
}

location @rubyproxy {
    proxy_connect_timeout 600;
    proxy_send_timeout 600;
    proxy_read_timeout 600;
    send_timeout 600;
    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_pass_header Server;
    proxy_pass http://0.0.0.0:3020;
}
```

This configuration allow you to use SEO cache only for bots (you can add other bots at the beginning of the file) and HTML GET requests. For connected users, you don't want to use the cached page (user connection, basket, ...). The performance are less important for users than bots.

## Heroku case

If you use Heroku server, you can't store files on dynos. But you have two alternatives:

- Use the memory mode

- Use a second server (a dedicated one) to store HTML files and combine with Nginx.

To intercept the request, use the following middleware in Rails:

In `config/initializers`, create a new file:

```ruby
require 'bot_detector'

if Rails.env.production?
  Rails.application.config.middleware.insert_before ActionDispatch::Static, BotDetector
end
``` 

Then in `lib` directory, for instance, manage the request:

```ruby
class BotRedirector
  CRAWLER_USER_AGENTS = ['googlebot', 'yahoo', 'bingbot', 'baiduspider', 'facebookexternalhit', 'twitterbot', 'rogerbot', 'linkedinbot', 'embedly', 'bufferbot', 'quora link preview', 'showyoubot', 'outbrain', 'pinterest/0.', 'developers.google.com/+/web/snippet', 'www.google.com/webmasters/tools/richsnippets', 'slackbot', 'vkShare', 'W3C_Validator', 'redditbot', 'Applebot', 'WhatsApp', 'flipboard', 'tumblr', 'bitlybot', 'SkypeUriPreview', 'nuzzel', 'Discordbot', 'Google Page Speed', 'Qwantify'].freeze

  IGNORE_URLS = [
    '/robots.txt'
  ].freeze

  def initialize(app)
    @app = app
  end

  def call(env)
    if env['HTTP_USER_AGENT'].present? && CRAWLER_USER_AGENTS.any? { |crawler_user_agent| env['HTTP_USER_AGENT'].downcase.include?(crawler_user_agent.downcase) }
      begin
        request = Rack::Request.new(env)

        return @app.call(env) if IGNORE_URLS.any? { |ignore_url| request.fullpath.downcase =~ /^#{ignore_url.downcase}/ }

        url     = URI.parse(ENV['SEO_SERVER'] + request.fullpath)
        headers = {
          'User-Agent'      => env['HTTP_USER_AGENT'],
          'Accept-Encoding' => 'gzip'
        }
        req     = Net::HTTP::Get.new(url.request_uri, headers)
        # req.basic_auth(ENV['SEO_USER_ID'], ENV['SEO_PASSWD']) # if authentication mechanism
        http         = Net::HTTP.new(url.host, url.port)
        http.use_ssl = true if url.scheme == 'https'
        response     = http.request(req)
        if response['Content-Encoding'] == 'gzip'
          response.body              = ActiveSupport::Gzip.decompress(response.body)
          response['Content-Length'] = response.body.length
          response.delete('Content-Encoding')
        end

        return [response.code.to_i, { 'Content-Type' => response.header['Content-Type'] }, [response.body]]
      rescue => error
        Rails.logger.error("[bot_redirection] #{error.message}")

        @app.call(env)
      end
    else
      @app.call(env)
    end
  end
end
```

If you use a second server, all links must be relatives in your HTML files, to avoid multi-domains links.

## Inspiration

Inspired by [prerender gem](https://github.com/prerender/prerender_rails).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/floXcoder/seo_cache. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the SeoCache project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/seo_cache/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
## 2.0.0

- Remove webdrivers dependency and use only selenium to fetch Chrome driver

## 1.4.0

- Correct Selenium warnings and add path for logging

## 1.3.0

- Add new option for Redis: redis_db_index

## 1.2.2

- Update Ruby version for Travis testing

## 1.2.1

- Update deprecated Selenium options

## 1.2.0

- Do not try to cache invalid requests

## 1.1.2

- Correct populate seo cache method

## 1.1.1

- Add new option to use locale as first directory for cache path

## 1.1.0

- Update dependencies

## 1.0.7

- Use new syntax to check if Redis key exists

## 1.0.6

- Logs only missed cache for success response

## 1.0.5

- Use correct path for index page with multiple domain names

## 1.0.4

- Add option to populate seo cache for multi-domains

## 1.0.3

- Ensure page rendered status is present after head element 

## 1.0.2

- Do not log missed cache for ignored statuses

## 1.0.1

- Add user agent to missed caches

## 1.0.0

- Update Readme (badges, syntax, links, ...)
- Add Travis CI
- Update dependencies
- Do not cache pages in error

## 0.19.0

- Add option to cache page if user connected

## 0.18.0

- Correct use of prerender param in URL

## 0.17.0

- Correct bug when populating cache

## 0.16.0

- Add prerender params when populating cache

## 0.15.0

- Add global option for Chrome debugging port problem

## 0.14.0

- Change options from Chrome
- Correct bug when rendering a 404 page

## 0.13.0

- Correct option for populate cache
- Add new option in Readme

## 0.12.0

- Force driver to quit after page rendering (except for populate cache, quit only after rendered all pages)
- Add option to wait for page loading (in case of asynchronous components, like React Router)

## 0.11.0

- Update for Rails 6

## 0.10.0

- Update Readme for nginx configuration

## 0.9.0

- Update gems
- Don't cache response if status is not 200
- Add 'seo_mode' variable to env to detect si seo mode is currently active
- Add status to source code to cache only pages with 200 HTTP status code
- Check existence of destination directory before creating it
- Remove persistent connection (already built-in in Selenium) 

## 0.8.0

- Change bundler version

## 0.7.0

- Add whitelist hosts domain option (authorize only these domains)

## 0.6.0

- Update gems (webdrivers and selenium-webdriver)

## 0.5.0

- Add more information in README
- Update gems and define cache time for browser driver

## 0.4.0

- Complete README
- Add new tests to cover disk and memory
- Remove "disk_" prefix from disk_cache_path and disk_cache_extension

## 0.3.0

- Add more examples in README
- Switch chromedriver to webdrivers
- Add tests
- Set correct URL for Github

## 0.2.0

- Improve README
- Add missing dependency

## 0.1.0

- First major release


spec/spec_helper.rb
# frozen_string_literal: true

require 'simplecov'
SimpleCov.start

require 'bundler/setup'
require 'rspec'
require 'webmock/rspec'
require 'seo_cache'

WebMock.disable_net_connect!(allow_localhost: true, allow: ['chromedriver.storage.googleapis.com'])

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  # config.disable_monkey_patching!

  config.expect_with :rspec do |conf|
    conf.syntax = :expect
  end
end


spec/seo_cache_spec.rb
# frozen_string_literal: true

describe SeoCache do
  it 'has a version number' do
    expect(SeoCache::VERSION).not_to be nil
  end
end


spec/seo_cache/page_render_spec.rb
# frozen_string_literal: true

require 'spec_helper'

require 'seo_cache/page_render'

describe SeoCache::PageRender do
  before(:all) do
    SeoCache.cache_mode            = 'memory'
    SeoCache.redis_namespace       = 'seo_cache_test'
    SeoCache.chrome_path           = '/usr/bin/chromium-browser'
    SeoCache.chrome_debugging_port = 9222
  end

  let(:page_render) { described_class }

  let(:redis) do
    uri = URI.parse(SeoCache.redis_url)
    Redis::Namespace.new(SeoCache.redis_namespace, redis: Redis.new(host: uri.host, port: uri.port, password: uri.password, db: SeoCache.redis_db_index, connect_timeout: 1, timeout: 1), warnings: false)
  end

  it 'renders a page' do
    rendered_page = page_render.new.get('https://example.com')
    expect(rendered_page).to match('Example Domain')
  end
end


lib/seo_cache/logger.rb
# frozen_string_literal: true

require 'forwardable'
require 'logger'

module SeoCache
  #
  # @example Enable full logging
  #   SeoCache.logger.level = :debug
  #
  # @example Log to file
  #   SeoCache.logger.output = 'seo_cache.log'
  #
  # @example Use logger manually
  #   SeoCache.logger.info('This is info message')
  #   SeoCache.logger.warn('This is warning message')
  #
  class Logger
    extend Forwardable
    include ::Logger::Severity

    def_delegators :@logger, :debug, :debug?,
                   :info, :info?,
                   :warn, :warn?,
                   :error, :error?,
                   :fatal, :fatal?,
                   :level

    def initialize(logger_path = nil)
      @logger = create_logger(logger_path || $stdout)
    end

    def output=(io)
      # `Logger#reopen` was added in Ruby 2.3
      if @logger.respond_to?(:reopen)
        @logger.reopen(io)
      else
        @logger = create_logger(io)
      end
    end

    #
    # For Ruby < 2.3 compatibility
    # Based on https://github.com/ruby/ruby/blob/ruby_2_3/lib/logger.rb#L250
    #

    def level=(severity)
      if severity.is_a?(Integer)
        @logger.level = severity
      else
        case severity.to_s.downcase
        when 'debug'
          @logger.level = DEBUG
        when 'info'
          @logger.level = INFO
        when 'warn'
          @logger.level = WARN
        when 'error'
          @logger.level = ERROR
        when 'fatal'
          @logger.level = FATAL
        when 'unknown'
          @logger.level = UNKNOWN
        else
          raise ArgumentError, "invalid log level: #{severity}"
        end
      end
    end

    #
    # Returns IO object used by logger internally.
    #
    # Normally, we would have never needed it, but we want to
    # use it as IO object for all child processes to ensure their
    # output is redirected there.
    #
    # It is only used in debug level, in other cases output is suppressed.
    #
    # @api private
    #
    def io
      @logger.instance_variable_get(:@logdev).instance_variable_get(:@dev)
    end

    #
    # Marks code as deprecated with replacement.
    #
    # @param [String] old
    # @param [String] new
    #
    def deprecate(old, new)
      warn "[DEPRECATION] #{old} is deprecated. Use #{new} instead."
    end

    private

    def create_logger(output)
      logger           = ::Logger.new(output)
      logger.progname  = 'SeoCache'
      logger.level     = ($DEBUG ? DEBUG : INFO)
      logger.formatter = proc do |severity, time, progname, msg|
        "#{time.strftime('%F %T')} #{severity} #{progname} #{msg}\n"
      end

      logger
    end
  end
end


lib/seo_cache/middleware.rb
# frozen_string_literal: true

require 'seo_cache/page_caching'
require 'seo_cache/page_render'

module SeoCache
  class Middleware
    def initialize(app, options = {})
      @options              = options
      @extensions_to_ignore = SeoCache.extensions_to_ignore
      @crawler_user_agents  = SeoCache.crawler_user_agents

      @app = app

      @page_caching = PageCaching.new
    end

    def call(env)
      @status, @headers, @response = @app.call(env)

      if prerender_page?(env, @status)
        cached_response = before_render(env)

        return cached_response.finish if cached_response.present?

        if SeoCache.prerender_service_url.present?
          prerender_response = prerender_service(env)
          if prerender_response
            response = build_response_from_prerender(prerender_response.body)
            after_render(env, prerender_response)

            return response.finish
          end
        else
          Thread.new do
            prerender_data = page_render(env)

            after_render(env, prerender_data, @status)
          end
        end
      elsif prerender_params?(env)
        env['seo_mode']           = true
        status, headers, response = @app.call(env)
        begin
          return [status, headers, [response.body]]
        rescue
          return [status, headers, [nil]]
        end
      end

      return [@status, @headers, @response]
    end

    def prerender_params?(env)
      return false if env['REQUEST_METHOD'] != 'GET'

      request      = Rack::Request.new(env)
      query_params = Rack::Utils.parse_query(request.query_string)

      return false if @extensions_to_ignore.any? { |extension| request.fullpath.include? extension }

      return true if query_params.has_key?(SeoCache.prerender_url_param) || query_params.has_key?(SeoCache.force_cache_url_param)
    end

    def prerender_page?(env, status)
      user_agent                   = env['HTTP_USER_AGENT']
      buffer_agent                 = env['HTTP_X_BUFFERBOT']
      is_requesting_prerender_page = false

      return false unless user_agent

      return false if status.to_i > 299

      return false if env['REQUEST_METHOD'] != 'GET'

      request      = Rack::Request.new(env)
      query_params = Rack::Utils.parse_query(request.query_string)

      # Check if user connected
      return false if !SeoCache.cache_with_user_connected && (env['warden'].respond_to?(:user) && env['warden'].user.present?)

      # If it is the generated page...don't prerender
      return false if query_params.has_key?(SeoCache.prerender_url_param)

      # if it is a bot and host doesn't contain these domains...don't prerender
      return false if SeoCache.whitelist_hosts.present? && SeoCache.whitelist_hosts.none? { |host| request.host.include?(host) }

      # if it is a bot and urls contain these params...don't prerender
      return false if SeoCache.blacklist_params.present? && SeoCache.blacklist_params.any? { |param| query_params.has_key?(param) }

      # if it is a bot and is requesting a resource...don't prerender
      return false if @extensions_to_ignore.any? { |extension| request.fullpath.include?(extension) }

      # if it is a bot and not requesting a resource and is not whitelisted...don't prerender
      return false if SeoCache.whitelist_urls.present? && SeoCache.whitelist_urls.all? { |whitelisted| !Regexp.new(whitelisted).match(request.fullpath) }

      # if it is a bot and not requesting a resource and is blacklisted(url or referer)...don't prerender
      blacklisted_url = SeoCache.blacklist_urls.present? && SeoCache.blacklist_urls.any? do |blacklisted|
        regex = Regexp.new(blacklisted)

        blacklisted_url     = regex.match(request.fullpath)
        blacklisted_referer = request.referer ? regex.match(request.referer) : false

        blacklisted_url || blacklisted_referer
      end
      return false if blacklisted_url

      is_requesting_prerender_page = true if Rack::Utils.parse_query(request.query_string).has_key?('_escaped_fragment_') || Rack::Utils.parse_query(request.query_string).has_key?(SeoCache.force_cache_url_param)

      # if it is a bot...show prerendered page
      is_requesting_prerender_page = true if @crawler_user_agents.any? { |crawler_user_agent| user_agent.downcase.include?(crawler_user_agent.downcase) }

      # if it is BufferBot...show prerendered page
      is_requesting_prerender_page = true if buffer_agent

      SeoCache.log('force cache : ' + request.path) if Rack::Utils.parse_query(request.query_string).has_key?(SeoCache.force_cache_url_param) && SeoCache.log_missed_cache

      return is_requesting_prerender_page
    end

    def prerender_service(env)
      url          = URI.parse(build_api_url(env))
      headers      = {
        'User-Agent'      => env['HTTP_USER_AGENT'],
        'Accept-Encoding' => 'gzip'
      }
      req          = Net::HTTP::Get.new(url.request_uri, headers)
      http         = Net::HTTP.new(url.host, url.port)
      http.use_ssl = true if url.scheme == 'https'
      response     = http.request(req)
      if response['Content-Encoding'] == 'gzip'
        response.body              = ActiveSupport::Gzip.decompress(response.body)
        response['Content-Length'] = response.body.length
        response.delete('Content-Encoding')
      end

      return response
    rescue StandardError => error
      SeoCache.log_error(error.message)
    end

    def build_api_url(env)
      new_env = env
      if env['CF-VISITOR']
        match = /"scheme":"(http|https)"/.match(env['CF-VISITOR'])
        (new_env['HTTPS'] = true) && (new_env['rack.url_scheme'] = 'https') && (new_env['SERVER_PORT'] = 443) if match && match[1] == 'https'
        (new_env['HTTPS'] = false) && (new_env['rack.url_scheme'] = 'http') && (new_env['SERVER_PORT'] = 80) if match && match[1] == 'http'
      end

      if env['X-FORWARDED-PROTO']
        (new_env['HTTPS'] = true) && (new_env['rack.url_scheme'] = 'https') && (new_env['SERVER_PORT'] = 443) if env['X-FORWARDED-PROTO'].split(',')[0] == 'https'
        (new_env['HTTPS'] = false) && (new_env['rack.url_scheme'] = 'http') && (new_env['SERVER_PORT'] = 80) if env['X-FORWARDED-PROTO'].split(',')[0] == 'http'
      end

      if SeoCache.protocol.present?
        (new_env['HTTPS'] = true) && (new_env['rack.url_scheme'] = 'https') && (new_env['SERVER_PORT'] = 443) if @options[:protocol] == 'https'
        (new_env['HTTPS'] = false) && (new_env['rack.url_scheme'] = 'http') && (new_env['SERVER_PORT'] = 80) if @options[:protocol] == 'http'
      end

      url           = Rack::Request.new(new_env).url
      prerender_url = SeoCache.prerender_service_url
      forward_slash = prerender_url[-1, 1] == '/' ? '' : '/'
      "#{prerender_url}#{forward_slash}#{url}"
    end

    def build_response_from_prerender(prerender_response)
      response = Rack::Response.new(prerender_response.body, prerender_response.code, prerender_response.header)

      # @options[:build_rack_response_from_prerender]&.call(response, prerendered_response)

      return response
    end

    def before_render(env)
      # return nil unless @options[:before_render]
      # cached_render = @options[:before_render].call(env)

      env_request = Rack::Request.new(env)

      cached_render = @page_caching.get(env_request.path, get_locale(env_request.params, env_request.host))

      return nil unless cached_render

      if cached_render.is_a?(String)
        Rack::Response.new(cached_render, 200, 'Content-Type' => 'text/html; charset=utf-8')
      elsif cached_render.is_a?(Rack::Response)
        cached_render
      end
    end

    def page_render(env)
      # Add key parameter to url
      request = Rack::Request.new(env)
      url     = if request.query_string.present? || request.url.include?('?')
                  request.url + '&'
                else
                  request.url + '?'
                end
      url     += "#{SeoCache.prerender_url_param}=true"

      return PageRender.new.get(url)
    end

    def after_render(env, response, status = 200)
      return unless response && SeoCache.cache_only_status.include?(status.to_i)

      env_request = Rack::Request.new(env)

      if SeoCache.log_missed_cache
        SeoCache.log("missed cache : #{env_request.path} (User agent: #{env_request.user_agent})")
      end

      @page_caching.cache(response, env_request.path, get_locale(env_request.params, env_request.host))
    end

    def get_locale(params, host)
      if SeoCache.locale_as_first_directory
        SeoCache.locale_method ? SeoCache.locale_method.call(params, host) : I18n.locale
      else
        nil
      end
    end
  end
end


lib/seo_cache/page_render.rb
# frozen_string_literal: true

module SeoCache
  class PageRender
    def initialize
      init_driver
    end

    def get(url, quit_after_render = true)
      @driver.get(url)

      sleep SeoCache.wait_time_for_page_loading if SeoCache.wait_time_for_page_loading

      return @driver.page_source
    rescue StandardError => error
      SeoCache.log_error(error.message)

      return false
    ensure
      @driver&.quit if quit_after_render
    end

    def close_connection
      @driver&.quit
    end

    private

    def init_driver
      if SeoCache.logger_path
        Selenium::WebDriver.logger.output = SeoCache.logger_path
        Selenium::WebDriver.logger.level  = SeoCache.logger_level
      end

      # Webdrivers.cache_time            = 86_400 * 10 # Chromedriver will be cached for 10 days (except if it detects a new version of Chrome)

      Selenium::WebDriver::Chrome.path = SeoCache.chrome_path if SeoCache.chrome_path

      browser_options = %w[headless incognito no-sandbox disable-gpu disable-infobars disable-dev-shm-usage disable-web-security disable-extensions disable-logging disable-notifications disable-sync window-size=1920x1080]
      browser_options << "remote-debugging-port=#{SeoCache.chrome_debugging_port}" if SeoCache.chrome_debugging_port

      @driver = ::Selenium::WebDriver.for(
        :chrome,
        options: Selenium::WebDriver::Chrome::Options.new(
          args: browser_options
        )
      )
    end
  end
end


lib/seo_cache/page_caching.rb
# frozen_string_literal: true

module SeoCache
  class PageCaching
    def initialize
      @redis = nil

      initialize_memory_cache if SeoCache.memory_cache?
    end

    def initialize_memory_cache
      uri    = URI.parse(SeoCache.redis_url)
      @redis = Redis::Namespace.new(SeoCache.redis_namespace, redis: Redis.new(host: uri.host, port: uri.port, password: uri.password, db: SeoCache.redis_db_index, connect_timeout: 1, timeout: 1), warnings: false)
    end

    def get(path, locale_domain = nil, extension = nil)
      @redis.get(cache_path(path, locale_domain, extension)) if SeoCache.memory_cache? && @redis
    end

    def cache(content, path, locale_domain = nil, extension = nil, gzip = Zlib::BEST_COMPRESSION)
      instrument :write_page, path do
        if SeoCache.memory_cache? && @redis
          write_to_memory(content, cache_path(path, locale_domain, extension))
        else
          write_to_disk(content, cache_path(path, locale_domain, extension), gzip)
        end
      end
    end

    def expire(path, locale_domain = nil, extension = nil)
      instrument :expire_page, path do
        delete(cache_path(path, locale_domain, extension))
      end
    end

    def cache_exists?(path, locale_domain = nil, extension = nil)
      if SeoCache.memory_cache? && @redis
        @redis.exists?(cache_path(path, locale_domain, extension))
      else
        File.exist?(cache_path(path, locale_domain, extension))
      end
    end

    private

    def cache_directory
      SeoCache.cache_path
    end

    def default_extension
      SeoCache.cache_extension
    end

    def cache_file(path, locale_domain, extension)
      name = if path.empty? || path =~ %r{\A/+\z}
               '/index'
             else
               URI::Parser.new.unescape(path.chomp('/'))
             end

      name = "#{locale_domain}/#{name}" if locale_domain

      name += extension || default_extension if File.extname(name).empty?

      name
    end

    def cache_path(path, locale_domain = nil, extension = nil)
      File.join(cache_directory, cache_file(path, locale_domain, extension))
    end

    def write_to_memory(content, path)
      @redis&.set(path, content)
    end

    def write_to_disk(content, path, gzip)
      FileUtils.makedirs(File.dirname(path)) unless File.directory?(File.dirname(path))
      File.open(path, 'wb+') { |f| f.write(content) }

      Zlib::GzipWriter.open(path + '.gz', gzip) { |f| f.write(content) } if gzip
    end

    def disk_delete(path)
      File.delete(path) if File.exist?(path)
      File.delete(path + '.gz') if File.exist?(path + '.gz')
    end

    def memory_delete(path)
      @redis.del(path) if @redis&.exists?(path)
    end

    def instrument(name, path)
      ActiveSupport::Notifications.instrument("#{name}.seo_cache", path: path) { yield }
    end
  end
end


lib/seo_cache.rb
# frozen_string_literal: true

require 'active_support'
require 'net/http'
require 'redis'
require 'redis-namespace'
require 'selenium/webdriver'

require 'seo_cache/logger'
require 'seo_cache/version'
require 'seo_cache/middleware'

module SeoCache

  mattr_accessor :chrome_path
  self.chrome_path = nil

  mattr_accessor :cache_mode # disk or memory
  self.cache_mode = 'memory'

  mattr_accessor :cache_path
  self.cache_path = ''

  mattr_accessor :locale_as_first_directory
  self.locale_as_first_directory = false

  mattr_accessor :locale_method
  self.locale_method = nil

  mattr_accessor :cache_extension
  self.cache_extension = '.html'

  mattr_accessor :redis_url
  self.redis_url = 'redis://localhost:6379/'

  mattr_accessor :redis_db_index
  self.redis_db_index = 0

  mattr_accessor :redis_namespace
  self.redis_namespace = '_seo_cache:'

  mattr_accessor :whitelist_hosts
  self.whitelist_hosts = []

  mattr_accessor :blacklist_urls
  self.blacklist_urls = []

  mattr_accessor :whitelist_urls
  self.whitelist_urls = []

  mattr_accessor :blacklist_params
  self.blacklist_params = []

  mattr_accessor :cache_with_user_connected
  self.cache_with_user_connected = false

  mattr_accessor :log_missed_cache
  self.log_missed_cache = false

  mattr_accessor :protocol
  self.protocol = nil

  mattr_accessor :prerender_service_url
  self.prerender_service_url = nil

  # Selenium doesn't support HTTP header, so URL parameter is mandatory
  mattr_accessor :prerender_url_param
  self.prerender_url_param = '_prerender_'

  mattr_accessor :force_cache_url_param
  self.force_cache_url_param = '_seo_cache_'

  # Time to wait to wait before page rendering (for asynchronous called), in seconds
  mattr_accessor :wait_time_for_page_loading
  self.wait_time_for_page_loading = nil

  mattr_accessor :cache_only_status
  self.cache_only_status = [200]

  # Default port used by Chrome: 9222
  mattr_accessor :chrome_debugging_port
  self.chrome_debugging_port = false

  mattr_accessor :extensions_to_ignore
  self.extensions_to_ignore = %w[.js .css .xml .less .png .jpg .jpeg .gif .pdf .doc .txt .ico .rss .zip .mp3 .rar .exe .wmv .doc .avi .ppt .mpg .mpeg .tif .wav .mov .psd .ai .xls .mp4 .m4a .swf .dat .dmg .iso .flv .m4v .torrent .woff2 .woff .gz .ttf .svg]

  mattr_accessor :crawler_user_agents
  self.crawler_user_agents = [
    'googlebot',
    'yahoo',
    'bingbot',
    'baiduspider',
    'facebookexternalhit',
    'twitterbot',
    'rogerbot',
    'linkedinbot',
    'embedly',
    'bufferbot',
    'quora link preview',
    'showyoubot',
    'outbrain',
    'pinterest/0.',
    'developers.google.com/+/web/snippet',
    'www.google.com/webmasters/tools/richsnippets',
    'slackbot',
    'vkShare',
    'W3C_Validator',
    'redditbot',
    'Applebot',
    'WhatsApp',
    'flipboard',
    'tumblr',
    'bitlybot',
    'SkypeUriPreview',
    'nuzzel',
    'Discordbot',
    'Google Page Speed',
    'Qwantify'
  ]

  mattr_accessor :logger_path
  self.logger_path = nil

  mattr_accessor :logger_level
  self.logger_level = :warn

  mattr_accessor :logger
  # self.logger = SeoCache::Logger.new(SeoCache.logger_path)

  def self.memory_cache?
    SeoCache.cache_mode == 'memory'
  end

  def self.disk_cache?
    SeoCache.cache_mode == 'disk'
  end

  def self.logger
    @logger ||= SeoCache::Logger.new(SeoCache.logger_path)
  end

  def self.log(message)
    SeoCache.logger.info(message)
    # Rails.logger.info { "[seo_cache] #{message}" }
  end

  def self.log_error(message)
    SeoCache.logger.error(message)
    # Rails.logger.error { "[seo_cache] #{message}" }
  end
end
